zerver/views/registration.py

```diff
@@ -338,9 +339,12 @@ def accounts_register(
             realm_name = form.cleaned_data["realm_name"]
             realm_type = form.cleaned_data["realm_type"]
             is_demo_org = form.cleaned_data["is_demo_organization"]
-            realm = do_create_realm(
-                string_id, realm_name, org_type=realm_type, is_demo_organization=is_demo_org
-            )
+            try:
+                realm = do_create_realm(
+                    string_id, realm_name, org_type=realm_type, is_demo_organization=is_demo_org
+                )
+            except IntegrityError as e:
+                raise JsonableError(_("Organization already exists")) from e
```

Comment:
I'm trying to think about whether we can do better with the user experience here. I think that if you pick a subdomain that's already taken, you do get an error earlier, so the most likely situation here is an actual race, which is 90%+ likely to be a race with yourself (i.e. you managed to double-submit the form). In that case, I wonder whether it'd be better to just serve a redirect to the organization rather than throwing a user-facing error. Thoughts? 

I haven't tried to actually reproduce this manually, but I'd guess one could by putting a `time.sleep(5)` in the right place.

Reply:
I could not manually create the error case. There is a very early check to ensure the realm doesn't exist or assert, so it would have to be two very close executions where the "first" realm hasn't been created yet when the "second" gets past that check. There is a lot of plumbing going on in the _accounts_register_ function that perhaps should get torn out so that things could occur more atomically, but that's not for this :smile:

Reply:
I can replicate with a `time.sleep(5)` on line 150 here:
https://github.com/zulip/zulip/blob/0cf836119b81854b0566ed24072aadc9fd556150/zerver/actions/create_realm.py#L144-L180

The race is between the check on line 144 and the creation on line 180.  It's a hard race to hit unless you put a time delay in there, or, as Tim calls out, someone double-clicks.

I agree that `accounts_register` is pretty gnarly.  I think to do what Tim is suggesting, we'd have to load up the already-created realm if we hit an integrity error, and return that from `do_create_realm`.  But then we're in the _exact same_ sort of race, except now with the creation of the user (it's created in an atomic transaction, but we can fail because we'd violate a unique index).  Which is, I guess, another type of race that the user-creation codepath should I guess deal with more gracefully?  It'd be the same sort of fix, of catching the integrity error and returning the already-created object, I think.

Reply:
As a extremely minor side-note, the `from e` in:
```py3
raise JsonableError(_("Organization already exists")) from e
```
isn't necessary if it's in an `except` block -- that bit [happens automatically](https://docs.python.org/3/tutorial/errors.html#exception-chaining).

Reply:
Thanks for the tip, I've removed the superfluous bits!

Reply:
> But then we're in the exact same sort of race, except now with the creation of the user (it's created in an atomic transaction, but we can fail because we'd violate a unique index). 

I think trying to return the user that'd already created on an IntegrityError has some security risk -- we'd have to be careful to avoid logging you in as a user account you'd tried to create.

And the other option, of serving a redirect to the app on IntegrityError (basically assuming you've created the account successfully with the request that won the race) has its own risks. 

So maybe the right plan is to just go with an exception like this, and if we want to eliminate the user-facing error in this unlikely race, address that with some JavaScript to make the button unclickable or similar.

(The user will get an email indicating their account creation worked with the URL, so it's not irrecoverable if they get this error).